v2/pkg/engine/resolve/resolve.go

```diff
@@ -642,7 +661,9 @@ func (r *Resolver) handleRemoveSubscription(id SubscriptionIdentifier) {
 				if ctx.Context().Err() == nil {
 					s.writer.Complete()
 				}
+				r.heartbeatSubLock.Lock()
 				delete(r.heartbeatSubscriptions, ctx)
+				r.heartbeatSubLock.Unlock()
```

Comment:
I personally would wrap these kind of blocks into a functions like `safeDeleteXYZ` 

e.g.
```
func safe(){
  r.lock()
  defer r.unlock() // make sure to always unlock
  ... function logic
}
```